pubtools/_quay/push_docker.py

```diff
@@ -538,49 +539,53 @@ def run(self):
-                existing_index_images = operator_pusher.get_existing_index_images(
-                    self.dest_quay_client
-                )
-                iib_results = operator_pusher.build_index_images()
-                # Sign operator images
-                operator_signature_handler.sign_operator_images(iib_results, index_stamp)
-                # Push index images to Quay
-                operator_pusher.push_index_images(iib_results, index_stamp)
-        except (Exception, SystemExit):
-            LOG.error("An exception has occurred during the push, starting rollback")
-            self.rollback(backup_tags, rollback_tags)
-            raise
-        else:
-            # Remove old signatures
-            self.remove_old_signatures(
-                docker_push_items,
-                operator_push_items,
-                existing_index_images,
-                iib_results,
-                backup_tags,
-                container_signature_handler,
-                operator_signature_handler,
-                sig_remover,
+        # Sign container images
+        container_signature_handler = ContainerSignatureHandler(
+            self.hub, self.task_id, self.target_settings, self.target_name
+        )
+        operator_signature_handler = OperatorSignatureHandler(
+            self.hub, self.task_id, self.target_settings, self.target_name
+        )
+        sig_remover = SignatureRemover()
+        sig_remover.set_quay_client(self.dest_quay_client)
+
+        container_signature_handler.sign_container_images(docker_push_items)
+        # Push container images
+        container_pusher = ContainerImagePusher(docker_push_items, self.target_settings)
+        container_pusher.push_container_images()
+
+        if operator_push_items:
+            # Build index images
+            operator_pusher = OperatorPusher(operator_push_items, self.target_settings)
+            existing_index_images = operator_pusher.get_existing_index_images(self.dest_quay_client)
+            iib_results = operator_pusher.build_index_images()
+            # Sign operator images
+            successful_results = dict(
+                [(key, val) for key, val in iib_results.items() if val["iib_result"]]
             )
+            operator_signature_handler.sign_operator_images(successful_results, index_stamp)
+            # Push index images to Quay
+            operator_pusher.push_index_images(successful_results, index_stamp)
+            # Rollback only when all index image builds fails
+            if set([x["iib_result"] for x in iib_results.values()]) == set([False]):
+                LOG.error("Push of all index images failed, running rollback.")
+                self.rollback(backup_tags, rollback_tags)
+                sys.exit(1)
+            if successful_results != iib_results:
+                LOG.error("Push of some index images failed")
+                sys.exit(1)
```

Comment:
I am just wondering if it is safe to use `sys.exit()` instead of `return` or `raise`?

Reply:
Is there a reason for using sys.exit?

Reply:
we catch for SystemExit in pub

Reply:
Could you elaborate on that?

Reply:
in EntryPointBase there's code in run method:
except SystemExit as e:
...

Reply:
But is there a particular reason for this solution? What would happen if you just raised a normal error?